File: apps/ios/Pebbles/Features/Glyph/Utils/PathSimplification.swift
// swiftlint:disable identifier_name
import CoreGraphics

/// Ramer-Douglas-Peucker path simplification.
/// Direct port of `apps/web/lib/utils/simplify-path.ts` (ε = 1.5 in callers).
enum PathSimplification {
    static func simplify(points: [CGPoint], epsilon: Double) -> [CGPoint] {
        guard points.count > 2 else { return points }

        var maxDist = 0.0
        var maxIndex = 0
        let first = points[0]
        let last = points[points.count - 1]

        for i in 1..<(points.count - 1) {
            let dist = perpendicularDistance(points[i], lineStart: first, lineEnd: last)
            if dist > maxDist {
                maxDist = dist
                maxIndex = i
            }
        }

        if maxDist > epsilon {
            let left = simplify(points: Array(points[0...maxIndex]), epsilon: epsilon)
            let right = simplify(points: Array(points[maxIndex..<points.count]), epsilon: epsilon)
            return Array(left.dropLast()) + right
        }

        return [first, last]
    }

    private static func perpendicularDistance(
        _ point: CGPoint,
        lineStart: CGPoint,
        lineEnd: CGPoint
    ) -> Double {
        let dx = lineEnd.x - lineStart.x
        let dy = lineEnd.y - lineStart.y
        let lengthSq = Double(dx * dx + dy * dy)

        if lengthSq == 0 {
            let ex = Double(point.x - lineStart.x)
            let ey = Double(point.y - lineStart.y)
            return (ex * ex + ey * ey).squareRoot()
        }

        let t = max(
            0.0,
            min(
                1.0,
                Double((point.x - lineStart.x) * dx + (point.y - lineStart.y) * dy) / lengthSq
            )
        )
        let projX = Double(lineStart.x) + t * Double(dx)
        let projY = Double(lineStart.y) + t * Double(dy)
        let ex = Double(point.x) - projX
        let ey = Double(point.y) - projY
        return (ex * ex + ey * ey).squareRoot()
    }
}
// swiftlint:enable identifier_name

File: apps/android/app/src/main/kotlin/app/pbbls/android/features/glyph/carve/PathSimplification.kt
package app.pbbls.android.features.glyph.carve

import kotlin.math.sqrt

/**
 * A point in carve space — Double math end to end, mirroring the iOS
 * `CGPoint`-based pipeline so simplification decisions can't drift.
 */
data class CarvePoint(
    val x: Double,
    val y: Double,
)

/**
 * Ramer–Douglas–Peucker simplification — ports iOS `PathSimplification.swift`
 * byte-for-byte in behavior (M43 design D1):
 *
 * - distance is to the **clamped segment** (t ∈ [0,1]), not the infinite line;
 * - the split is strict (`maxDist > epsilon`);
 * - the max-distance point lands in BOTH recursion halves, joined
 *   `left.dropLast(1) + right`;
 * - `size <= 2` (including empty) passes through unchanged.
 *
 * Epsilon is applied in canvas space (280dp side) BEFORE the ×(200/side)
 * viewBox scale — simplifying after scaling changes the output.
 */
object PathSimplification {
    fun simplify(
        points: List<CarvePoint>,
        epsilon: Double,
    ): List<CarvePoint> {
        if (points.size <= 2) return points
        val first = points.first()
        val last = points.last()
        var maxDist = 0.0
        var maxIndex = 0
        for (i in 1 until points.size - 1) {
            val dist = perpendicularDistance(points[i], first, last)
            if (dist > maxDist) {
                maxDist = dist
                maxIndex = i
            }
        }
        return if (maxDist > epsilon) {
            val left = simplify(points.subList(0, maxIndex + 1), epsilon)
            val right = simplify(points.subList(maxIndex, points.size), epsilon)
            left.dropLast(1) + right
        } else {
            listOf(first, last)
        }
    }

    /** Distance from [point] to the segment [lineStart]→[lineEnd], t clamped to [0,1]. */
    private fun perpendicularDistance(
        point: CarvePoint,
        lineStart: CarvePoint,
        lineEnd: CarvePoint,
    ): Double {
        val dx = lineEnd.x - lineStart.x
        val dy = lineEnd.y - lineStart.y
        val lengthSq = dx * dx + dy * dy
        if (lengthSq == 0.0) {
            return euclidean(point, lineStart)
        }
        val t = (((point.x - lineStart.x) * dx + (point.y - lineStart.y) * dy) / lengthSq).coerceIn(0.0, 1.0)
        val projected = CarvePoint(lineStart.x + t * dx, lineStart.y + t * dy)
        return euclidean(point, projected)
    }

    private fun euclidean(
        a: CarvePoint,
        b: CarvePoint,
    ): Double {
        val dx = a.x - b.x
        val dy = a.y - b.y
        return sqrt(dx * dx + dy * dy)
    }
}
